hotel_restaurant/models/hotel_restaurant.py

```diff
@@ -0,0 +1,699 @@
+            self.partner_address_id = addr["default"]
+
+    @api.onchange("folio_id")
+    def _onchange_folio_id(self):
+        """
+        When you change folio_id, based on that it will update
+        the customer_id and room_number as well
+        ---------------------------------------------------------
+        @param self: object pointer
+        """
+        for rec in self:
+            rec.write({"customer_id": False, "room_id": False})
+            if rec.folio_id:
+                rec.customer_id = rec.folio_id.partner_id.id
+                if rec.folio_id.room_lines:
+                    rec.room_id = rec.folio_id.room_lines[0].product_id.id
+
+    def action_set_to_draft(self):
+        """
+        This method is used to change the state
+        to draft of the hotel restaurant reservation
+        --------------------------------------------
+        @param self: object pointer
+        """
+        self.write({"state": "draft"})
+
+    def table_reserved(self):
+        """
+        when CONFIRM BUTTON is clicked this method is called for
+        table reservation
+        @param self: The object pointer
+        @return: change a state depending on the condition
+        """
+        for reservation in self:
+            self._cr.execute(
+                "select count(*) from "
+                "hotel_restaurant_reservation as hrr "
+                "inner join reservation_table as rt on \
+                             rt.reservation_table_id = hrr.id "
+                "where (start_date,end_date)overlaps\
+                             ( timestamp %s , timestamp %s ) "
+                "and hrr.id<> %s and state != 'done'"
+                "and rt.name in (select rt.name from \
+                             hotel_restaurant_reservation as hrr "
+                "inner join reservation_table as rt on \
+                             rt.reservation_table_id = hrr.id "
+                "where hrr.id= %s) ",
+                (
+                    reservation.start_date,
+                    reservation.end_date,
+                    reservation.id,
+                    reservation.id,
+                ),
+            )
+            res = self._cr.fetchone()
+            roomcount = res and res[0] or 0.0
+            if len(reservation.table_nos_ids.ids) == 0:
+                raise ValidationError(
+                    _(
+                        "Please Select Tables For \
+                                         Reservation"
+                    )
+                )
+            if roomcount:
+                raise ValidationError(
+                    _(
+                        "You tried to confirm reservation \
+                                         with table those already reserved \
+                                         in this reservation period"
+                    )
+                )
+            self.state = "confirm"
+            return True
```

Comment:
Either use the reservation inside the loop instead of self or use self.ensure_one()
and return statement should be outside of the loop